Feature/integration test

cmprescott opened the pull request:
Thanks for merging my last pull request.

I was planning on expanding this role out so I can use it on my macbook also. I figured I should also implement CI to make sure I wasn't breaking anything. Turned out to be more changes than I thought. Travis is pretty fussy about file permissions in /home/travis/.config/... 

Finally got it working by making more commands sudo than I thought I would have to, but end result seems to be the same :-)

igormukhingmailcom commented on defaults/main.yml:
```diff
@@ -2,6 +2,8 @@
 
 # --- Defaults (Welcome to override!) ---
 packagecontrol_url: "https://packagecontrol.io/Package Control.sublime-package"
+packagecontrol_dir_bin: "~/.config/sublime-text-3/Installed\ Packages"
```
Not sure that slash needed inside quotes here, in README.md and tasks/main.yml.
WDYT?

cmprescott replied:
Good catch. See https://github.com/cmprescott/ansible-role-sublimetext3_packagecontrol/commit/f8f935219dfe7c2d16f90a5aff742fc6a5e7b229 

This was just one of the things I tried to get Travis to complete the build. However, it is not needed.

igormukhingmailcom commented:
First of all, thank you again for your contribution.

I recommend `jdauphant.sublimetext` in `README.md` to use with my role and it will be not pretty good if in tests will be another role. I'll prefer to recommend yours if you added it to galaxy because it seems it more flexible.

I see that you make a lot of changes in your fork of `jdauphant.sublimetext` role, so there are 2 different ways:
- Wait for merging https://github.com/jdauphant/ansible-role-sublimetext/pull/3 and use `jdauphant.sublimetext` in `test.yml` (prefer way, but PR merging not guaranteed)
- Make your own role with version tag to use it in `test.yml` and `README.md` (you don't like to initiate competing role, but it seems to be good solution I think)

WDYT?

cmprescott commented:
After thinking about it some more I created my own Ansible Galaxy Role. I prefer the role name spell out it's for Sublime Text 3 instead of having it be a note. I don't think there is any easy way for original person to update that without potentially breaking people's scripts. 

Anyways, the galaxy role still passes the integration test :+1: 

Thanks for the feedback on my pull request! It's nice to have a second set of eyes.

cmprescott commented:
PS -- If all is good merge back has some TODOs that you will have to do
- TODO: Add original project to TravisCI
- TODO: Change url on build badge from cmprescott to yours

igormukhingmailcom commented on .travis.yml:
```diff
@@ -0,0 +1,12 @@
+  - pip install ansible==1.8.0
+script:
+  - ansible-galaxy install -r tests/requirements.yml
+  - ansible-playbook -i tests/inventory --syntax-check --list-tasks tests/test.yml -e "role_name=ansible-role-sublimetext3_packagecontrol" -e "hosts_group=hosts_group"
```
Can you please tell what does it mean:
`-e "role_name=ansible-role-sublimetext3_packagecontrol" -e "hosts_group=hosts_group"`
Have same result output when I cut that.

Looks like rudiment from another role that have `tests/test.yml` with next content:

``` yml
---
- hosts: "{{ hosts_group }}"
  roles:
    - role: "{{ role_name }}"
```

cmprescott replied:
Yep that is lazy copy pasting on my part. I originally half stole my ideas for travis code from https://github.com/jdauphant/ansible-role-nginx and https://github.com/RHInception/ansible-xml 

I was new to Travis and had a lot of the 'DAMN YOU TRAVIS!' commits. I also want to say when I ran an older version of Ansible it would throw an error if `role_name` !== the role's name. I tried it out on 1.8.0 and that doesn't seem to be the case anymore. 

The hosts_group is 100% uneccessary.

Thanks for catching that! I have that code in way too many of my ansible roles. I need to go an updated that. Even worse I submitted a pull request to someone with that code and he has been copy pasting it too. I'll be sure to let him know and credit you with for pointing out the mistake. 

Thanks,
Chris

igormukhingmailcom commented on .travis.yml:
```diff
@@ -0,0 +1,12 @@
+ - sudo apt-get update -qq
+ - sudo apt-get install -qq python-apt python-pycurl
+install:
+  - pip install ansible==1.8.0
```
Just for information. 
This is minimal version that role can run on travis?
If you don't tested lower version - I'll do this after merge and sync with `meta/main.yml` (by the way I'm not sure my role work with 1.4 as I have only 1.9 on my computer).

cmprescott replied:
Experimentally the dependent (my sublime text 3) role requires ansible version 1.8.0+. See https://github.com/cmprescott/ansible-role-sublimetext3/commit/b758e1c1fb9f56ab7cd0428d0820a11ac6b049a9 where I tried 1.4.4 and 1.7.0. 

I also talked about the 1.8.0 requirement with original sublimetext role author https://github.com/jdauphant/ansible-role-sublimetext/pull/3#issuecomment-90984850

I think your role should work on 1.4+. If that's the case you can use `sudo apt-get install ansible` instead of `pip install ansible==1.8.0`. This speeds up the Travis build.

Let me know your thoughts.

igormukhingmailcom commented on tasks/main.yml:
```diff
@@ -2,21 +2,33 @@
     owner: "{{ packagecontrol_owner }}"
     group: "{{ packagecontrol_group }}"
+  with_items:
+    - "{{ packagecontrol_dir_bin | expanduser }}"
```
I think this must be discussed a little bit.

I noticed when tested role that `expanuser` expands only local user. Running this role on remote machine (e.g. vagrant box) have unexpected (as for me) behavior.

I asked about this here: https://github.com/ansible/ansible/issues/10670#issuecomment-92358611 and realize that this is not a bug.

On the one hand, this role seems to be used only locally (for me). On the other hand, that breaks tests on vagrant box or make unable to install that role on remote machine.

WDYT?

cmprescott committed:
Add integration tests and travis config

- Uses local connection
- Installs dependent role via ansible-galaxy requirements file
- Installs with defaults and settings from readme

cmprescott committed:
Don't track test roles

- Helps keep SCM clean when working locally

cmprescott committed:
Fixes to get Travis to play nice.

- Add Travis build bag to see results easier
- Ensure the dirs before files, so installs correctly even when sublime hasn't created the dir yet

cmprescott committed:
Try lower version of Ansible

- I believe 1.6 is lowest possible. 1.6 needed for deb option of apt module http://docs.ansible.com/apt_module.html

cmprescott committed:
Set travis back to using 1.8 ansible

- 1.6.4 gets "ERROR: cannot find role in /home/travis/build/cmprescott/ansible-role-sublimetext3_packagecontrol/tests/roles/jdauphant.sublimetext or /home/travis/build/cmprescott/ansible-role-sublimetext3_packagecontrol/tests/jdauphant.sublimetext or /etc/ansible/roles/jdauphant.sublimetext" when trying to install dependencies from ansible galaxy.

cmprescott commented:
@igormukhingmailcom thought of this the other day. Rewrote history on my feature branch. Code is significantly less hacked now. Trick was package control dir(s) needed user +x to avoid file permission issues. Also, it uses jdauphant's role again instead of mine. He merged the changes I wanted, so I deleted mine.

igormukhingmailcom commented:
Hi, @cmprescott!
Thank you a lot for your contribution.

I came to the fact that it is not necessary to test external role here, only the local one. So I had to remove related code.
